src/backend/distributed/transaction/transaction_management.c

```diff
@@ -317,12 +319,19 @@ CoordinatedTransactionCallback(XactEvent event, void *arg)
 			MemoryContext previousContext =
 				MemoryContextSwitchTo(CitusXactCallbackContext);
 
-			if (CurrentCoordinatedTransactionState == COORD_TRANS_PREPARED)
+			if (CurrentCoordinatedTransactionState == COORD_TRANS_PREPARED &&
+				!IsMainDBCommand)
 			{
 				/* handles both already prepared and open transactions */
 				CoordinatedRemoteTransactionsCommit();
 			}
 
+			if (!IsCitusMainDB())
```

Comment:
I'm still concerned about executing below UDF even for regular Citus clusters, due to the check (if !IsCitusMainDb()) not knowing about whether this database is used for db sharding or row-based sharding. Should we check for IsCitusInstalled() too? Or am I missing something?

Reply:
I think if we're in a sharded not-main database the ideal behaviour would be to create the roles via the current database but add the pg_dist_object records via the main database. (Both as part of the same 2PC)

Until we implement that, we should make sure we keep whatever the current behaviour is when there are multiple Citus databases.

Reply:
> IsCitusInstalled()

Can use CitusHasBeenLoaded probably